nexus/src/app/sagas/instance_create.rs

```diff
@@ -60,174 +50,286 @@ pub struct Params {
+            "AllocatePropolisIp",
+            ALLOC_PROPOLIS_IP.as_ref(),
+        ));
+
+        builder.append(Node::action(
+            "instance_name",
+            "CreateInstanceRecord",
+            CREATE_INSTANCE_RECORD.as_ref(),
+        ));
+
+        // Helper function for append subsagas to our parent saga.
+        fn subsaga_append<S: Serialize>(
+            node_basename: &'static str,
+            subsaga_builder: steno::DagBuilder,
+            parent_builder: &mut steno::DagBuilder,
+            params: S,
+            which: usize,
+        ) -> Result<(), SagaInitError> {
+            // The "parameter" node is a constant node that goes into the outer
+            // saga.  Its value becomes the parameters for the one-node subsaga
+            // (defined below) that actually creates each NIC.
+            let params_node_name = format!("{}_params{}", node_basename, which);
+            parent_builder.append(Node::constant(
+                &params_node_name,
+                serde_json::to_value(&params).map_err(|e| {
+                    SagaInitError::SerializeError(params_node_name.clone(), e)
+                })?,
+            ));
 
-    // Similar to disks (see block comment below), create a sequence of
-    // action-undo pairs for each requested external IP, to make sure we always
-    // unwind after failures.
-    for i in 0..MAX_EXTERNAL_IPS_PER_INSTANCE {
-        template_builder.append(
-            format!("external_ip_id{i}").as_str(),
-            "CreateExternalIpId",
-            new_action_noop_undo(saga_generate_uuid),
-        );
+            let output_name = format!("{}{}", node_basename, which);
+            parent_builder.append(Node::subsaga(
+                output_name.as_str(),
+                subsaga_builder.build()?,
+                params_node_name,
+            ));
+            Ok(())
+        }
 
-        template_builder.append(
-            format!("external_ip{i}").as_str(),
-            "CreateExternalIp",
-            ActionFunc::new_action(
-                async move |sagactx| {
-                    sic_allocate_instance_external_ip(sagactx, i).await
-                },
-                async move |sagactx| {
-                    sic_allocate_instance_external_ip_undo(sagactx, i).await
-                },
-            ),
-        );
-    }
+        // We use a similar pattern here for NICs, external IPs and disks.  We
+        // want one saga action per item (i.e., per NIC, per external IP, or per
+        // disk).  That makes it much easier to make actions and undo actions
+        // idempotent.  But the user may ask for a variable number of these
+        // items.  Previous versions of Steno required the saga DAG to be fixed
+        // for all runs of a saga.  To address this, we put a static limit on
+        // the number of NICs, external IPs, or disks that you can request.
+        // Here where we're building the saga DAG, we always add that maximum
```

Comment:
Grammar: `Here, `

Reply:
Fixed in 33951e28e39cfd0dfeacf5066700dff6f206b045.